### PR TITLE
linsolve: add BiCG and BiCGStab methods

### PR DESCRIPTION
**Pull request by vladimir-ch**
Testing this is as difficult as any other iterative algorithm: a correct result does not guarantee that the implementation is correct, and a failing test does not mean that the implementation is wrong. I had to tweak a bit but I'd like to move forward now.

Please take a look.

### CODE REVIEW

File: linsolve/bcgstab.go [hunk 1]
```diff
@@ -0,0 +1,144 @@
...
+		}
+		fallthrough
+	case 5:
+		// Solve M^{-1} * r_i.
+		copy(ctx.Src, ctx.Residual)
+		b.resume = 6
+		return PreconSolve, nil
+	case 6:
+		copy(b.shat, ctx.Dst)
+		// Compute A * s^_i.
```
Comment (btracey) [inline linsolve/bcgstab.go]: \hat{s}_i

File: linsolve/bcgstab.go [hunk 2]
```diff
@@ -0,0 +1,144 @@
...
+		copy(b.t, ctx.Dst)
+		b.omega = floats.Dot(b.t, ctx.Residual) / floats.Dot(b.t, b.t)
+		floats.AddScaled(ctx.X, b.omega, b.shat)
+		floats.AddScaled(ctx.Residual, -b.omega, b.t)
+		b.resume = 8
+		return CheckResidual, nil
+	case 8:
+		if !ctx.Converged && math.Abs(b.omega) < omegaBreakdownTol {
+			b.resume = 0
+			return NoOperation, errors.New("bcgstab: omega breakdown")
```
Comment (btracey) [inline linsolve/bcgstab.go]: Maybe a `Breakdown` error with some more information pieces?

File: linsolve/bcg.go [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+// preconditioning for solving systems of linear equations
+//  A*x = b,
+// where A is a square, generally nonsymmetric matrix.
+//
+// References:
+//  - Barrett, R. et al. (1994). Section 2.3.5 BiConjugate Gradient (BiCG).
+//    In Templates for the Solution of Linear Systems: Building Blocks
+//    for Iterative Methods (2nd ed.) (pp. 19-20). Philadelphia, PA: SIAM.
+//    Retrieved from http://www.netlib.org/templates/templates.pdf
+type BCG struct {
```
[the remark this answers is not in the document]
Comment (vladimir-ch) [reply linsolve/bcg.go]: Done.

File: linsolve/bcgstab.go [hunk 4]
```diff
@@ -0,0 +1,144 @@
...
+//  A*x = b,
+// where A is a square, generally nonsymmetric matrix.
+//
+//
+// References:
+//  - Barrett, R. et al. (1994). Section 2.3.8 BiConjugate Gradient Stabilized (Bi-CGSTAB).
+//    In Templates for the Solution of Linear Systems: Building Blocks
+//    for Iterative Methods (2nd ed.) (pp. 24-25). Philadelphia, PA: SIAM.
+//    Retrieved from http://www.netlib.org/templates/templates.pdf
+type BCGSTAB struct {
```
[the remark this answers is not in the document]
Comment (vladimir-ch) [reply linsolve/bcgstab.go]: Done.

File: linsolve/bcgstab.go [hunk 5]
```diff
@@ -0,0 +1,144 @@
...
+		}
+		b.alpha = b.rho / rtv
+		// Form the residual and X so that we can check for tolerance early.
+		floats.AddScaled(ctx.X, b.alpha, b.phat)
+		floats.AddScaled(ctx.Residual, -b.alpha, b.v)
+		b.resume = 4
+		return CheckResidual, nil
+	case 4:
+		if ctx.Converged {
+			b.resume = 5
```
[the remark this answers is not in the document]
Comment (vladimir-ch) [reply linsolve/bcgstab.go]: It's good you brought this up. I was split between forcing the user to call Init after a convergence or being benevolent and allowing him to continue the iteration for whatever reason they might have. Using `resume = 0` will also simplify the implementation at a few places. Done.

File: linsolve/bcgstab.go (hunk 2 above)
Comment (vladimir-ch) [reply linsolve/bcgstab.go]: Added `BreakdownError` with `Value` and `Tolerance` fields, not sure if that's useful or if something else should be added.

File: linsolve/bcgstab.go [hunk 6]
```diff
@@ -0,0 +1,144 @@
...
+func (b *BCGSTAB) Iterate(ctx *Context) (Operation, error) {
+	switch b.resume {
+	case 1:
+		if b.first {
+			copy(b.rt, ctx.Residual)
+		}
+		b.rho = floats.Dot(b.rt, ctx.Residual)
+		if math.Abs(b.rho) < rhoBreakdownTol {
+			b.resume = 0
+			return NoOperation, errors.New("bcgstab: rho breakdown")
```
[the remark this answers is not in the document]
Comment (vladimir-ch) [reply linsolve/bcgstab.go]: Yes, I combined them into a single `BreakdownError`, it's probably not necessary to distinguish between different causes of a breakdown.

File: linsolve/bcg.go [hunk 7]
```diff
@@ -0,0 +1,120 @@
...
+		// Solve M^{-T} * rt_{i-1}.
+		copy(ctx.Src, b.rt)
+		b.resume = 3
+		return PreconSolve | Trans, nil
+	case 3:
+		copy(b.zt, ctx.Dst)
+		b.rho = floats.Dot(b.z, b.rt)
+		if math.Abs(b.rho) < rhoBreakdownTol {
+			b.resume = 0
+			return NoOperation, errors.New("bcg: rho breakdown")
```
[the remark this answers is not in the document]
Comment (vladimir-ch) [reply linsolve/bcg.go]: Added a new error type `BreakdownError`.

**Comment by vladimir-ch** [comment]
PTAL

**Comment by vladimir-ch** [comment]
_ping_

**Comment by vladimir-ch** [comment]
Sorry for pinging again but could you please take a look @btracey ? I have a bit of spare time to make some further progress on linsolve.

**Comment by btracey** [comment]
I'll try to get to it tonight. Could you look at the optimize PR?

**Comment by vladimir-ch** [comment]
Thanks. I'll take a look at the optimize PR soon.

**Comment by btracey** [comment]
I'm most of the way through a CL to have the tests pass

**Pull request merged**